File: frontend-app/src/app/services/modulo-reporte.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { ModuloReporte } from "../domain/ModuloReporte";


@Injectable({
    providedIn: 'root'
})
export class ModuloReporteService{

    private url:string = 'http://localhost:8080/api/modulo-reporte';

    constructor(private http: HttpClient){

    }

    listadoModuloReportes(): Observable<ModuloReporte[]>{
        return this.http.get<ModuloReporte[]>(this.url);
    }
    
        
    buscarModuloReportePorId(id: number): Observable<ModuloReporte>{
        return this.http.get<ModuloReporte>(`${this.url}/${id}`);
    }
    
    guardarModuloReporte(moduloReporte: ModuloReporte){
        return this.http.post('http://localhost:8080/api/modulo-reporte/guardar', moduloReporte);
    }
        
    editarModuloReporte(moduloReporte: ModuloReporte) {
        return this.http.put(`${this.url}/actualizar`, moduloReporte);
    }
        
    eliminarModuloReporte(idModuloReporte: number){
        return this.http.delete(`${this.url}/${idModuloReporte}`);
    }

    listadoModuloReportePaginacion(page: number): Observable<any>{
        return this.http.get<any[]>(`${this.url}/page/${page}`);
    }   

    buscarModuloReportePorNombre(nombreModuloReporte: string): Observable<ModuloReporte[]>{
        const params = new HttpParams().set('nombreModuloReporte', nombreModuloReporte);
                        
        return this.http.get<ModuloReporte[]>(`${this.url}/buscar`, { params });
    }

    listadoModulosReportesCompletos(): Observable<ModuloReporte[]>{
        return this.http.get<ModuloReporte[]>(`${this.url}/completos`);
    }

}

File: backend-app/src/main/java/com/backend_app/backend_app/service/ModuloReporteService.java
package com.backend_app.backend_app.service;

import com.backend_app.backend_app.domain.ModuloReporte;
import com.backend_app.backend_app.dto.ModuloReporteDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface ModuloReporteService {

    List<ModuloReporteDTO> listadoModuloReportes();

    Optional<ModuloReporteDTO> buscarModuloReportePorId(long idModulo);

    ModuloReporteDTO convertirADTO(ModuloReporte moduloReporte);

    void guardarModuloReporte(ModuloReporteDTO moduloReporteDTO);

    void actualizarModuloReporte(ModuloReporteDTO moduloReporteDTO);

    void eliminarModuloReporte(ModuloReporteDTO moduloReporteDTO);

    Page<ModuloReporteDTO> findAll(Pageable pageable);

    List<ModuloReporteDTO> buscarModuloReportePorNombre(String nombreModuloReporte);

    List<ModuloReporteDTO> listadoModulosReportesCompletos();

}
